lib/elixir/lib/module/checker.ex

```diff
@@ -68,42 +68,60 @@ defmodule Module.Checker do
 
   defp check_remote(module, fun, arity, meta, state) do
     cond do
-      not warn_if_undefined?(module, fun, arity, state) ->
-        []
-
       # TODO: In the future we may want to warn for modules defined
       # in the local context
       Keyword.get(meta, :context_module, false) and state.module != module ->
         []
 
-      # TODO: Add no_autoload
-      not Code.ensure_loaded?(module) ->
+      not Code.ensure_loaded?(module) and warn_undefined?(module, fun, arity, state) ->
         warn(meta, state, {:undefined_module, module, fun, arity})
 
-      not function_exported?(module, fun, arity) ->
+      not function_exported?(module, fun, arity) and warn_undefined?(module, fun, arity, state) ->
         exports = exports_for(module)
         warn(meta, state, {:undefined_function, module, fun, arity, exports})
 
+      (reason = deprecated(module, fun, arity)) && warn_deprecated?(module, fun, arity, state) ->
+        warn(meta, state, {:deprecated, module, fun, arity, reason})
+
       true ->
         []
     end
   end
 
   # TODO: Do not warn inside guards
   # TODO: Properly handle protocols
-  defp warn_if_undefined?(_module, :__impl__, 1, _state), do: false
-  defp warn_if_undefined?(:erlang, :orelse, 2, _state), do: false
-  defp warn_if_undefined?(:erlang, :andalso, 2, _state), do: false
+  defp warn_undefined?(_module, :__impl__, 1, _state), do: false
+  defp warn_undefined?(:erlang, :orelse, 2, _state), do: false
+  defp warn_undefined?(:erlang, :andalso, 2, _state), do: false
+
+  defp warn_undefined?(module, fun, arity, state) do
+    warn_compile_opts?(module, fun, arity, state.compile_opts, :no_warn_undefined)
```

Comment:
Would it maybe make sense to build a `MapSet` out of the option and use that for the lookup? It seems to me it should be both faster and a bit simpler implementation-wise.

Reply:
This is only invoked once an undefined function is found. So it is unlikely to happen.